File: client-clojure/src/sirqul_io_t_platform/specs/application_response.clj
(ns sirqul-io-t-platform.specs.application-response
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.name-string-value-response :refer :all]
            [sirqul-io-t-platform.specs.asset-short-response :refer :all]
            [sirqul-io-t-platform.specs.asset-short-response :refer :all]
            [sirqul-io-t-platform.specs.platform-response :refer :all]
            [sirqul-io-t-platform.specs.category-response :refer :all]
            [sirqul-io-t-platform.specs.billable-entity-short-response :refer :all]
            [sirqul-io-t-platform.specs.trilat-app-settings :refer :all]
            )
  (:import (java.io File)))


(def application-response-data
  {
   (ds/opt :valid) boolean?
   (ds/opt :message) string?
   (ds/opt :version) float?
   (ds/opt :serializeNulls) boolean?
   (ds/opt :startTimeLog) int?
   (ds/opt :errorCode) string?
   (ds/opt :request) (s/coll-of name-string-value-response-spec)
   (ds/opt :applicationId) int?
   (ds/opt :appKey) string?
   (ds/opt :title) string?
   (ds/opt :currentBuildVersion) string?
   (ds/opt :currentApiVersion) string?
   (ds/opt :appIcon) asset-short-response-spec
   (ds/opt :appLogo) asset-short-response-spec
   (ds/opt :active) boolean?
   (ds/opt :about) string?
   (ds/opt :landingPageUrl) string?
   (ds/opt :landingPageFullUrl) string?
   (ds/opt :defaultImageFullUrl) string?
   (ds/opt :activityDescription) string?
   (ds/opt :platforms) (s/coll-of platform-response-spec)
   (ds/opt :categories) (s/coll-of category-response-spec)
   (ds/opt :placementCount) int?
   (ds/opt :billable) billable-entity-short-response-spec
   (ds/opt :hasAds) boolean?
   (ds/opt :publicNotifications) boolean?
   (ds/opt :urlScheme) string?
   (ds/opt :global) boolean?
   (ds/opt :downloadUrls) string?
   (ds/opt :localAdsEnabled) boolean?
   (ds/opt :trilatSettings) trilat-app-settings-spec
   (ds/opt :appType) string?
   (ds/opt :appName) string?
   (ds/opt :bundleId) string?
   (ds/opt :scoringType) string?
   (ds/opt :hintCost) int?
   (ds/opt :maxScore) int?
   (ds/opt :ticketsPerPoint) float?
   (ds/opt :hasGameData) boolean?
   (ds/opt :globalTickets) boolean?
   (ds/opt :eulaVersion) string?
   (ds/opt :eulaDateUpdated) int?
   (ds/opt :modules) string?
   (ds/opt :billableEntityName) string?
   (ds/opt :showInActivities) boolean?
   (ds/opt :cssClass) string?
   (ds/opt :inviteWelcomeText) string?
   (ds/opt :invitePageUrl) string?
   (ds/opt :apnsCertificateFileName) string?
   (ds/opt :facebookAppId) string?
   (ds/opt :facebookAppSecret) string?
   (ds/opt :appToken) string?
   (ds/opt :googleApiKey) string?
   (ds/opt :appRestKey) string?
   (ds/opt :publicContentApproval) boolean?
   (ds/opt :productionMode) boolean?
   (ds/opt :minimumSessionLength) int?
   (ds/opt :sessionGapLength) int?
   (ds/opt :sqootApiKey) string?
   (ds/opt :amqpEnabled) boolean?
   (ds/opt :authorizedCount) int?
   (ds/opt :authorizedServers) string?
   (ds/opt :returning) string?
   })

(def application-response-spec
  (ds/spec
    {:name ::application-response
     :spec application-response-data}))

File: client-scalaz/src/main/scala/org/openapitools/client/api/ApplicationResponse.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import ApplicationResponse._

case class ApplicationResponse (
  valid: Option[Boolean],
message: Option[String],
version: Option[Double],
serializeNulls: Option[Boolean],
startTimeLog: Option[Long],
errorCode: Option[String],
request: Option[List[NameStringValueResponse]],
applicationId: Option[Long],
appKey: Option[String],
title: Option[String],
currentBuildVersion: Option[String],
currentApiVersion: Option[String],
appIcon: Option[AssetShortResponse],
appLogo: Option[AssetShortResponse],
active: Option[Boolean],
about: Option[String],
landingPageUrl: Option[String],
landingPageFullUrl: Option[String],
defaultImageFullUrl: Option[String],
activityDescription: Option[String],
platforms: Option[List[PlatformResponse]],
categories: Option[List[CategoryResponse]],
placementCount: Option[Integer],
billable: Option[BillableEntityShortResponse],
hasAds: Option[Boolean],
publicNotifications: Option[Boolean],
urlScheme: Option[String],
global: Option[Boolean],
downloadUrls: Option[String],
localAdsEnabled: Option[Boolean],
trilatSettings: Option[TrilatAppSettings],
appType: Option[String],
appName: Option[String],
bundleId: Option[String],
scoringType: Option[ScoringType],
hintCost: Option[Integer],
maxScore: Option[Integer],
ticketsPerPoint: Option[Float],
hasGameData: Option[Boolean],
globalTickets: Option[Boolean],
eulaVersion: Option[String],
eulaDateUpdated: Option[Long],
modules: Option[String],
billableEntityName: Option[String],
showInActivities: Option[Boolean],
cssClass: Option[String],
inviteWelcomeText: Option[String],
invitePageUrl: Option[String],
apnsCertificateFileName: Option[String],
facebookAppId: Option[String],
facebookAppSecret: Option[String],
appToken: Option[String],
googleApiKey: Option[String],
appRestKey: Option[String],
publicContentApproval: Option[Boolean],
productionMode: Option[Boolean],
minimumSessionLength: Option[Integer],
sessionGapLength: Option[Integer],
sqootApiKey: Option[String],
amqpEnabled: Option[Boolean],
authorizedCount: Option[Integer],
authorizedServers: Option[String],
returning: Option[String])

object ApplicationResponse {
  import DateTimeCodecs._
  sealed trait ScoringType
  case object GAMELEVEL extends ScoringType
  case object GAMEOBJECT extends ScoringType

  object ScoringType {
    def toScoringType(s: String): Option[ScoringType] = s match {
      case "GAMELEVEL" => Some(GAMELEVEL)
      case "GAMEOBJECT" => Some(GAMEOBJECT)
      case _ => None
    }

    def fromScoringType(x: ScoringType): String = x match {
      case GAMELEVEL => "GAMELEVEL"
      case GAMEOBJECT => "GAMEOBJECT"
    }
  }

  implicit val ScoringTypeEnumEncoder: EncodeJson[ScoringType] =
    EncodeJson[ScoringType](is => StringEncodeJson(ScoringType.fromScoringType(is)))

  implicit val ScoringTypeEnumDecoder: DecodeJson[ScoringType] =
    DecodeJson.optionDecoder[ScoringType](n => n.string.flatMap(jStr => ScoringType.toScoringType(jStr)), "ScoringType failed to de-serialize")

  implicit val ApplicationResponseCodecJson: CodecJson[ApplicationResponse] = CodecJson.derive[ApplicationResponse]
  implicit val ApplicationResponseDecoder: EntityDecoder[ApplicationResponse] = jsonOf[ApplicationResponse]
  implicit val ApplicationResponseEncoder: EntityEncoder[ApplicationResponse] = jsonEncoderOf[ApplicationResponse]
}
